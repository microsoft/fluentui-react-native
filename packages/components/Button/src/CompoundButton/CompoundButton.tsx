/** @jsx withSlots */
import * as React from 'react';
import { Pressable, View } from 'react-native';
import { ActivityIndicator } from '@fluentui-react-native/experimental-activity-indicator';
import { CompoundButtonProps, compoundButtonName, CompoundButtonType } from './CompoundButton.types';
import { TextV1 as Text } from '@fluentui-react-native/text';
import { stylingSettings } from './CompoundButton.styling';
import { compose, mergeProps, withSlots, UseSlots } from '@fluentui-react-native/framework';
import { useButton } from '../useButton';
import { Icon } from '@fluentui-react-native/icon';
import { createIconProps } from '@fluentui-react-native/interactive-hooks';
import { buttonLookup } from '../Button';

export const CompoundButton = compose<CompoundButtonType>({
  displayName: compoundButtonName,
  ...stylingSettings,
  slots: {
    root: Pressable,
    icon: Icon,
    content: Text,
    secondaryContent: Text,
    contentContainer: View,
  },
  useRender: (userProps: CompoundButtonProps, useSlots: UseSlots<CompoundButtonType>) => {
    const button = useButton(userProps);
    const iconProps = createIconProps(userProps.icon);

    // grab the styled slots
    const Slots = useSlots(userProps, (layer) => buttonLookup(layer, button.state, userProps));

    // now return the handler for finishing render
    return (final: CompoundButtonProps, ...children: React.ReactNode[]) => {
      const { icon, iconOnly, secondaryContent, iconPosition, loading, accessibilityLabel, ...mergedProps } = mergeProps(
        button.props,
        final,
      );

      const shouldShowIcon = !loading && icon;
      if (__DEV__ && iconOnly) {
        React.Children.forEach(children, (child) => {
          if (typeof child === 'string') {
            console.warn('iconOnly should not be set when Button has content.');
          }
        });
      }

      let childText = '';
      if (accessibilityLabel === undefined) {
        React.Children.forEach(children, (child) => {
          if (typeof child === 'string') {
            childText = child; // We only automatically support the one child string.
          }
        });

        if (secondaryContent) {
          childText += ' ' + secondaryContent;
        }
      }
      const label = accessibilityLabel ?? childText;

      return (
        <Slots.root {...mergedProps} accessibilityLabel={label}>
          {loading && <ActivityIndicator />}
          {shouldShowIcon && iconPosition === 'before' && <Slots.icon {...iconProps} />}
          <Slots.contentContainer>
            {React.Children.map(children, (child) =>
              typeof child === 'string' ? <Slots.content key="content">{child}</Slots.content> : child,
            )}
            {secondaryContent && <Slots.secondaryContent key="secondaryContent">{secondaryContent}</Slots.secondaryContent>}
          </Slots.contentContainer>
          {shouldShowIcon && iconPosition === 'after' && <Slots.icon {...iconProps} />}
        </Slots.root>
      );
    };
  },
});
