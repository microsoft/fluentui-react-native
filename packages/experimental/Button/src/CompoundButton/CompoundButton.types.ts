import { ViewProps, ColorValue } from 'react-native';
import { TextProps } from '@fluentui-react-native/experimental-text';
import { ButtonSlotProps, ButtonTokens, ButtonProps } from '../Button.types';
import { FontTokens } from '@fluentui-react-native/tokens';

export const compoundButtonName = 'CompoundButton';

export interface CompoundButtonTokens extends ButtonTokens {
  secondaryContentFont?: FontTokens;
  secondaryContentColor?: ColorValue;
}

export interface CompoundButtonProps extends ButtonProps {
  /**
   * Second line of text that describes the action this button takes.
   */
  secondaryContent?: string;
}

export interface CompoundButtonSlotProps extends ButtonSlotProps {
  contentContainer: ViewProps;
  secondaryContent: TextProps;
}

export interface CompoundButtonType {
  props: CompoundButtonProps;
  tokens: CompoundButtonTokens;
  slotProps: CompoundButtonSlotProps;
}
