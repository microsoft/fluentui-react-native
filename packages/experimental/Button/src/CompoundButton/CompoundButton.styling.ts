import { compoundButtonName, CompoundButtonTokens, CompoundButtonSlotProps, CompoundButtonProps } from './CompoundButton.types';
import { Theme, UseStylingOptions, buildProps } from '@fluentui-react-native/framework';
import { borderStyles, fontStyles, layoutStyles } from '@fluentui-react-native/tokens';
import { buttonStates, defaultButtonTokens } from '../ButtonTokens';
import { ButtonTokens } from '../Button.types';

export const stylingSettings: UseStylingOptions<CompoundButtonProps, CompoundButtonSlotProps, CompoundButtonTokens> = {
  tokens: [
    defaultButtonTokens,
    (t: Theme) => ({
      minHeight: 72,
      secondaryContentFont: {
        variant: 'secondaryStandard',
      },
      ghost: {
        secondaryContentColor: t.colors.ghostSecondaryContent,
        hovered: {
          secondaryContentColor: t.colors.ghostHoveredSecondaryContent,
        },
        focused: {
          secondaryContentColor: t.colors.ghostFocusedSecondaryContent,
        },
        pressed: {
          secondaryContentColor: t.colors.ghostPressedSecondaryContent,
        },
      } as ButtonTokens,
    }),
    compoundButtonName,
  ],
  states: buttonStates,
  slotProps: {
    root: buildProps(
      (tokens: CompoundButtonTokens, theme: Theme) => ({
        style: {
          display: 'flex',
          alignItems: 'center',
          flexDirection: 'row',
          alignSelf: 'flex-start',
          justifyContent: 'center',
          paddingStart: 16,
          paddingEnd: 16,
          backgroundColor: tokens.backgroundColor,
          ...borderStyles.from(tokens, theme),
          ...layoutStyles.from(tokens, theme),
        },
      }),
      ['backgroundColor', ...borderStyles.keys, ...layoutStyles.keys],
    ),
    contentContainer: {
      style: {
        display: 'flex',
        flexDirection: 'column',
      },
    },
    content: buildProps(
      (tokens: CompoundButtonTokens, theme: Theme) => ({
        style: {
          color: tokens.color,
          ...fontStyles.from(tokens, theme),
        },
      }),
      ['color', ...fontStyles.keys],
    ),
    secondaryContent: buildProps(
      (tokens: CompoundButtonTokens, theme: Theme) => ({
        style: {
          color: tokens.secondaryContentColor,
          ...fontStyles.from(tokens.secondaryContentFont, theme),
        },
      }),
      ['secondaryContentColor', 'secondaryContentFont'],
    ),
    icon: buildProps(
      (tokens: CompoundButtonTokens) => ({
        style: {
          tintColor: tokens.iconColor,
        },
      }),
      ['iconColor'],
    ),
  },
};
