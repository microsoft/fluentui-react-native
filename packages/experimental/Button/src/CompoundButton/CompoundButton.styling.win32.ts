import { compoundButtonName, CompoundButtonTokens, CompoundButtonSlotProps, CompoundButtonProps } from './CompoundButton.types';
import { Theme, UseStylingOptions, buildProps } from '@fluentui-react-native/framework';
import { borderStyles, fontStyles, layoutStyles } from '@fluentui-react-native/tokens';
import { defaultButtonTokens } from '../ButtonTokens';
import { ButtonTokens } from '../Button.types';
import { buttonStates } from '../Button.styling';

export const stylingSettings: UseStylingOptions<CompoundButtonProps, CompoundButtonSlotProps, CompoundButtonTokens> = {
  tokens: [
    defaultButtonTokens,
    (t: Theme): ButtonTokens =>
      ({
        minHeight: 72,
        secondaryContentFont: {
          variant: 'secondaryStandard',
        },
        secondaryContentColor: t.colors.neutralForeground2,
        hovered: {
          secondaryContentColor: t.colors.neutralForeground2Hover,
        },
        focused: {
          secondaryContentColor: t.colors.neutralForeground2Hover,
        },
        pressed: {
          secondaryContentColor: t.colors.neutralForeground2Pressed,
        },
        primary: {
          secondaryContentColor: t.colors.neutralForegroundOnBrand,
          hovered: {
            secondaryContentColor: t.colors.neutralForegroundOnBrand,
          },
          focused: {
            secondaryContentColor: t.colors.neutralForegroundOnBrand,
          },
          pressed: {
            secondaryContentColor: t.colors.neutralForegroundOnBrand,
          },
        },
        subtle: {
          secondaryContentColor: t.colors.neutralForeground2,
          hovered: {
            secondaryContentColor: t.colors.neutralForeground2Hover,
          },
          focused: {
            secondaryContentColor: t.colors.neutralForeground2Hover,
          },
          pressed: {
            secondaryContentColor: t.colors.neutralForeground2Pressed,
          },
        },
      } as ButtonTokens),
    compoundButtonName,
  ],
  states: buttonStates,
  slotProps: {
    root: buildProps(
      (tokens: CompoundButtonTokens, theme: Theme) => ({
        style: {
          display: 'flex',
          alignItems: 'center',
          flexDirection: 'row',
          alignSelf: 'flex-start',
          justifyContent: 'center',
          paddingStart: 16,
          paddingEnd: 16,
          backgroundColor: tokens.backgroundColor,
          ...borderStyles.from(tokens, theme),
          ...layoutStyles.from(tokens, theme),
        },
      }),
      ['backgroundColor', ...borderStyles.keys, ...layoutStyles.keys],
    ),
    contentContainer: {
      style: {
        display: 'flex',
        flexDirection: 'column',
      },
    },
    content: buildProps(
      (tokens: CompoundButtonTokens, theme: Theme) => ({
        style: {
          color: tokens.color,
          ...fontStyles.from(tokens, theme),
        },
      }),
      ['color', ...fontStyles.keys],
    ),
    secondaryContent: buildProps(
      (tokens: CompoundButtonTokens, theme: Theme) => ({
        style: {
          color: tokens.secondaryContentColor,
          ...fontStyles.from(tokens.secondaryContentFont, theme),
        },
      }),
      ['secondaryContentColor', 'secondaryContentFont'],
    ),
    icon: buildProps(
      (tokens: CompoundButtonTokens) => ({
        style: {
          tintColor: tokens.iconColor,
        },
      }),
      ['iconColor'],
    ),
  },
};
