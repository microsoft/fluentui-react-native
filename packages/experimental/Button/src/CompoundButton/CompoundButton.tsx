/** @jsx withSlots */
import * as React from 'react';
import { Image, View } from 'react-native';
import { CompoundButtonProps, compoundButtonName, CompoundButtonType } from './CompoundButton.types';
import { Text } from '@fluentui-react-native/experimental-text';
import { stylingSettings } from './CompoundButton.styling';
import { filterImageProps } from '@fluentui-react-native/adapters';
import { compose, mergeProps, withSlots, UseSlots } from '@fluentui-react-native/framework';
import { useButton } from '../useButton';

export const CompoundButton = compose<CompoundButtonType>({
  displayName: compoundButtonName,
  ...stylingSettings,
  slots: {
    root: View,
    icon: Image,
    content: Text,
    secondaryContent: Text,
    contentContainer: View,
  },
  filters: {
    icon: filterImageProps,
  },
  render: (userProps: CompoundButtonProps, useSlots: UseSlots<CompoundButtonType>) => {
    const button = useButton(userProps);

    // grab the styled slots
    const Slots = useSlots(userProps, (layer) => button.state[layer] || userProps[layer]);

    // now return the handler for finishing render
    return (final: CompoundButtonProps, ...children: React.ReactNode[]) => {
      const { icon, content, secondaryContent, ...mergedProps } = mergeProps(button.props, final);

      return (
        <Slots.root {...mergedProps}>
          {icon && <Slots.icon key="icon" source={{ uri: icon }} />}
          {(content || secondaryContent) && (
            <Slots.contentContainer>
              {content && <Slots.content key="content">{content}</Slots.content>}
              {secondaryContent && <Slots.secondaryContent key="secondaryContent">{secondaryContent}</Slots.secondaryContent>}
            </Slots.contentContainer>
          )}
          {children}
        </Slots.root>
      );
    };
  },
});

export default CompoundButton;
