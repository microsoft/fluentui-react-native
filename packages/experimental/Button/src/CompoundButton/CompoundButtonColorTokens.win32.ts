import { Theme } from '@fluentui-react-native/framework';
import { TokenSettings } from '@fluentui-react-native/use-styling';
import { CompoundButtonTokens } from './CompoundButton.types';

export const defaultCompoundButtonColorTokens: TokenSettings<CompoundButtonTokens, Theme> = (t: Theme): CompoundButtonTokens => ({
  secondaryContentColor: t.colors.neutralForeground2,
  hovered: {
    secondaryContentColor: t.colors.neutralForeground2Hover,
  },
  focused: {
    secondaryContentColor: t.colors.neutralForeground2Hover,
  },
  pressed: {
    secondaryContentColor: t.colors.neutralForeground2Pressed,
  },
  primary: {
    secondaryContentColor: t.colors.neutralForegroundOnBrand,
    hovered: {
      secondaryContentColor: t.colors.neutralForegroundOnBrand,
    },
    focused: {
      secondaryContentColor: t.colors.neutralForegroundOnBrand,
    },
    pressed: {
      secondaryContentColor: t.colors.neutralForegroundOnBrand,
    },
  },
  subtle: {
    secondaryContentColor: t.colors.neutralForeground2,
    hovered: {
      secondaryContentColor: t.colors.neutralForeground2Hover,
    },
    focused: {
      secondaryContentColor: t.colors.neutralForeground2Hover,
    },
    pressed: {
      secondaryContentColor: t.colors.neutralForeground2Pressed,
    },
  },
});
