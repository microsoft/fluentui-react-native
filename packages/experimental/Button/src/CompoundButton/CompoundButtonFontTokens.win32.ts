import { FontWeightValue, Theme } from '@fluentui-react-native/framework';
import { globalTokens } from '@fluentui-react-native/theme-tokens';
import { TokenSettings } from '@fluentui-react-native/use-styling';
import { CompoundButtonTokens } from './CompoundButton.types';

export const defaultCompoundButtonFontTokens: TokenSettings<CompoundButtonTokens, Theme> = (t: Theme): CompoundButtonTokens => ({
  medium: {
    hasContent: {
      fontFamily: t.typography.families.secondary,
      fontSize: globalTokens.font.size[200],
      fontWeight: globalTokens.font.weight.semibold as FontWeightValue,
      secondaryContentFont: {
        fontFamily: t.typography.families.secondary,
        fontSize: globalTokens.font.size[100],
        fontWeight: globalTokens.font.weight.semibold as FontWeightValue,
      },
    },
  },
  small: {
    hasContent: {
      fontFamily: t.typography.families.primary,
      fontSize: globalTokens.font.size[200],
      fontWeight: globalTokens.font.weight.regular as FontWeightValue,
      secondaryContentFont: {
        fontFamily: t.typography.families.secondary,
        fontSize: globalTokens.font.size[100],
        fontWeight: globalTokens.font.weight.semibold as FontWeightValue,
      },
    },
  },
  large: {
    hasContent: {
      fontFamily: t.typography.families.secondary,
      fontSize: globalTokens.font.size[400],
      fontWeight: globalTokens.font.weight.semibold as FontWeightValue,
      secondaryContentFont: {
        fontFamily: t.typography.families.primary,
        fontSize: globalTokens.font.size[200],
        fontWeight: globalTokens.font.weight.regular as FontWeightValue,
      },
    },
  },
});
