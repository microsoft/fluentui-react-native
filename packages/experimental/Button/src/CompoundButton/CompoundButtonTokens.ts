import { Theme } from '@fluentui-react-native/framework';
import { globalTokens } from '@fluentui-react-native/theme-tokens';
import { TokenSettings } from '@fluentui-react-native/use-styling';
import { CompoundButtonTokens } from './CompoundButton.types';

export const defaultCompoundButtonTokens: TokenSettings<CompoundButtonTokens, Theme> = (): CompoundButtonTokens => ({
  medium: {
    padding: globalTokens.spacing.m - globalTokens.stroke.width.thin,
    focused: {
      padding: globalTokens.spacing.m,
    },
    hasContent: {
      paddingHorizontal: globalTokens.spacing.m - globalTokens.stroke.width.thin,
      minWidth: 96,
      variant: 'bodySemibold',
      secondaryContentFont: { variant: 'secondaryStandard' },
      focused: {
        paddingHorizontal: globalTokens.spacing.m,
      },
      hasIconAfter: {
        spacingIconContentAfter: globalTokens.spacing.m,
      },
      hasIconBefore: {
        spacingIconContentBefore: globalTokens.spacing.m,
      },
    },
  },
  small: {
    padding: globalTokens.spacing.s - globalTokens.stroke.width.thin,
    focused: {
      padding: globalTokens.spacing.s,
    },
    hasContent: {
      paddingHorizontal: globalTokens.spacing.s - globalTokens.stroke.width.thin,
      minWidth: 64,
      variant: 'bodyStandard',
      secondaryContentFont: { variant: 'secondaryStandard' },
      focused: {
        paddingHorizontal: globalTokens.spacing.s,
      },
      hasIconAfter: {
        spacingIconContentAfter: globalTokens.spacing.s,
      },
      hasIconBefore: {
        spacingIconContentBefore: globalTokens.spacing.s,
      },
    },
  },
  large: {
    padding: globalTokens.spacing.l - globalTokens.stroke.width.thin,
    focused: {
      padding: globalTokens.spacing.l,
    },
    hasContent: {
      paddingHorizontal: globalTokens.spacing.l - globalTokens.stroke.width.thin,
      minWidth: 96,
      variant: 'subheaderSemibold',
      secondaryContentFont: { variant: 'bodyStandard' },
      focused: {
        paddingHorizontal: globalTokens.spacing.l,
      },
      hasIconAfter: {
        spacingIconContentAfter: globalTokens.spacing.l,
      },
      hasIconBefore: {
        spacingIconContentBefore: globalTokens.spacing.l,
      },
    },
  },
});
