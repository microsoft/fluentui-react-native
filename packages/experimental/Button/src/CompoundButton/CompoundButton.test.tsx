import * as React from 'react';
import { CompoundButton } from './CompoundButton';
import * as renderer from 'react-test-renderer';

it('CompoundButton default', () => {
  const tree = renderer.create(<CompoundButton content="Default Button" secondaryContent="sublabel" />).toJSON();
  expect(tree).toMatchSnapshot();
});
