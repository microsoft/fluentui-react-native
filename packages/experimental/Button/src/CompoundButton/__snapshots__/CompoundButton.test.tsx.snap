// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompoundButton default 1`] = `
<View
  accessibilityLabel="Default Button"
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  iconPosition="before"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyUp={[Function]}
  onPress={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#f3f2f1",
      "borderColor": "#8a8886",
      "borderRadius": 4,
      "borderWidth": 1,
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minWidth": 96,
      "padding": 11,
      "paddingHorizontal": 11,
    }
  }
>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#323130",
          "fontFamily": "Segoe UI",
          "fontSize": 14,
          "fontWeight": "600",
          "margin": 0,
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
        }
      }
    >
      Default Button
    </Text>
    <Text
      style={
        Object {
          "fontFamily": "Segoe UI",
          "fontSize": 12,
          "fontWeight": "400",
          "margin": 0,
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
        }
      }
    >
      sublabel
    </Text>
  </View>
</View>
`;
